File: go-backend-interview-prep/11_system_design_in_go/04_cache_service_mini/main.py
"""Cache service with TTL and LRU eviction -- Python equivalent."""

import threading
import time
from collections import OrderedDict


class CacheEntry:
    def __init__(self, value, expires_at):
        self.value = value
        self.expires_at = expires_at


class Cache:
    def __init__(self, capacity):
        self.lock = threading.Lock()
        self.items = OrderedDict()  # key -> CacheEntry (OrderedDict tracks access order)
        self.capacity = capacity

        # Start periodic cleanup
        t = threading.Thread(target=self._cleanup_loop, daemon=True)
        t.start()

    def set(self, key, value, ttl_seconds):
        with self.lock:
            if key in self.items:
                # Update existing: remove and re-insert at end (most recent)
                del self.items[key]

            if len(self.items) >= self.capacity:
                self._evict_lru()

            self.items[key] = CacheEntry(value, time.time() + ttl_seconds)

    def get(self, key):
        with self.lock:
            if key not in self.items:
                return None, False

            entry = self.items[key]

            # Lazy expiration
            if time.time() > entry.expires_at:
                del self.items[key]
                return None, False

            # Move to end (most recently used)
            self.items.move_to_end(key)
            return entry.value, True

    def size(self):
        with self.lock:
            return len(self.items)

    def _evict_lru(self):
        """Remove least recently used entry. Caller must hold lock."""
        if self.items:
            key, _ = self.items.popitem(last=False)  # pop from front (oldest)
            print(f"  [evict] key={key!r} (LRU)")

    def _cleanup_loop(self):
        while True:
            time.sleep(1)
            with self.lock:
                now = time.time()
                expired = [k for k, e in self.items.items() if now > e.expires_at]
                for k in expired:
                    print(f"  [cleanup] expired key={k!r}")
                    del self.items[k]


# --- Demo ---

def main():
    cache = Cache(capacity=3)

    print("=== cache demo (capacity=3) ===\n")

    cache.set("user:1", "Alice", ttl_seconds=5)
    cache.set("user:2", "Bob", ttl_seconds=2)
    cache.set("user:3", "Charlie", ttl_seconds=5)
    print(f"set 3 entries, cache size: {cache.size()}\n")

    # Cache hit
    val, ok = cache.get("user:1")
    if ok:
        print(f"GET user:1 -> {val} (hit)")

    val, ok = cache.get("user:2")
    if ok:
        print(f"GET user:2 -> {val} (hit)")

    # Cache miss
    _, ok = cache.get("user:99")
    if not ok:
        print("GET user:99 -> (miss)")

    # Trigger LRU eviction
    print("\nadding user:4 (should evict LRU)...")
    cache.set("user:4", "Diana", ttl_seconds=5)
    print(f"cache size: {cache.size()}")

    _, ok = cache.get("user:3")
    if not ok:
        print("GET user:3 -> (miss, was evicted as LRU)")

    # Wait for TTL expiration
    print("\nwaiting 3s for user:2 TTL to expire...")
    time.sleep(3)

    _, ok = cache.get("user:2")
    if not ok:
        print("GET user:2 -> (miss, expired)")

    print(f"\nfinal cache size: {cache.size()}")
    print("\ndemo done")


if __name__ == "__main__":
    main()

File: go-backend-interview-prep/11_system_design_in_go/04_cache_service_mini/main.go
package main

import (
	"container/list"
	"fmt"
	"sync"
	"time"
)

// --- Cache Entry ---

type entry struct {
	key       string
	value     interface{}
	expiresAt time.Time
	element   *list.Element // pointer into LRU list
}

// --- Cache ---

type Cache struct {
	mu       sync.Mutex
	items    map[string]*entry
	order    *list.List // front = most recent, back = least recent
	capacity int
}

func NewCache(capacity int) *Cache {
	c := &Cache{
		items:    make(map[string]*entry),
		order:    list.New(),
		capacity: capacity,
	}
	// Periodic cleanup of expired entries
	go c.cleanupLoop(1 * time.Second)
	return c
}

func (c *Cache) Set(key string, value interface{}, ttl time.Duration) {
	c.mu.Lock()
	defer c.mu.Unlock()

	// Update existing entry
	if e, ok := c.items[key]; ok {
		e.value = value
		e.expiresAt = time.Now().Add(ttl)
		c.order.MoveToFront(e.element)
		return
	}

	// Evict if at capacity
	if len(c.items) >= c.capacity {
		c.evictLRU()
	}

	// Add new entry
	e := &entry{
		key:       key,
		value:     value,
		expiresAt: time.Now().Add(ttl),
	}
	e.element = c.order.PushFront(e)
	c.items[key] = e
}

func (c *Cache) Get(key string) (interface{}, bool) {
	c.mu.Lock()
	defer c.mu.Unlock()

	e, ok := c.items[key]
	if !ok {
		return nil, false
	}

	// Lazy expiration check
	if time.Now().After(e.expiresAt) {
		c.removeEntry(e)
		return nil, false
	}

	// Move to front (most recently used)
	c.order.MoveToFront(e.element)
	return e.value, true
}

func (c *Cache) Len() int {
	c.mu.Lock()
	defer c.mu.Unlock()
	return len(c.items)
}

// evictLRU removes the least recently used entry. Caller must hold lock.
func (c *Cache) evictLRU() {
	back := c.order.Back()
	if back == nil {
		return
	}
	e := back.Value.(*entry)
	fmt.Printf("  [evict] key=%q (LRU)\n", e.key)
	c.removeEntry(e)
}

// removeEntry deletes an entry from both map and list. Caller must hold lock.
func (c *Cache) removeEntry(e *entry) {
	c.order.Remove(e.element)
	delete(c.items, e.key)
}

// cleanupLoop periodically removes expired entries.
func (c *Cache) cleanupLoop(interval time.Duration) {
	ticker := time.NewTicker(interval)
	defer ticker.Stop()
	for range ticker.C {
		c.mu.Lock()
		now := time.Now()
		for key, e := range c.items {
			if now.After(e.expiresAt) {
				fmt.Printf("  [cleanup] expired key=%q\n", key)
				c.removeEntry(e)
			}
		}
		c.mu.Unlock()
	}
}

// --- Demo ---

func main() {
	cache := NewCache(3) // max 3 entries

	fmt.Println("=== cache demo (capacity=3) ===")

	// Set entries with different TTLs
	cache.Set("user:1", "Alice", 5*time.Second)
	cache.Set("user:2", "Bob", 2*time.Second)
	cache.Set("user:3", "Charlie", 5*time.Second)
	fmt.Printf("set 3 entries, cache size: %d\n\n", cache.Len())

	// Cache hit
	if v, ok := cache.Get("user:1"); ok {
		fmt.Printf("GET user:1 -> %v (hit)\n", v)
	}
	if v, ok := cache.Get("user:2"); ok {
		fmt.Printf("GET user:2 -> %v (hit)\n", v)
	}

	// Cache miss
	if _, ok := cache.Get("user:99"); !ok {
		fmt.Printf("GET user:99 -> (miss)\n")
	}

	// Trigger LRU eviction (add 4th entry, capacity is 3)
	fmt.Println("\nadding user:4 (should evict LRU)...")
	cache.Set("user:4", "Diana", 5*time.Second)
	fmt.Printf("cache size: %d\n", cache.Len())

	// user:3 should be evicted (least recently accessed)
	if _, ok := cache.Get("user:3"); !ok {
		fmt.Printf("GET user:3 -> (miss, was evicted as LRU)\n")
	}

	// Wait for TTL expiration
	fmt.Println("\nwaiting 3s for user:2 TTL to expire...")
	time.Sleep(3 * time.Second)

	if _, ok := cache.Get("user:2"); !ok {
		fmt.Printf("GET user:2 -> (miss, expired)\n")
	}

	fmt.Printf("\nfinal cache size: %d\n", cache.Len())
	fmt.Println("\ndemo done")
}
